Vic2ToHoI4/Source/HOI4World/Map/Railways.cpp

```diff
@@ -0,0 +1,150 @@
+#include "Railways.h"
+#include "Log.h"
+
+
+
+using HoI4::Railway;
+
+
+namespace
+{
+
+std::shared_ptr<Vic2::Province> getValidVic2Province(int provinceNum,
+	 const std::map<int, std::shared_ptr<Vic2::Province>>& Vic2Provinces)
+{
+	const auto& itr = Vic2Provinces.find(provinceNum);
+	if (itr == Vic2Provinces.end())
+	{
+		return nullptr;
+	}
+	const auto& Vic2NeighborProvince = itr->second;
+	if (!Vic2NeighborProvince->isLandProvince())
+	{
+		return nullptr;
+	}
+
+	return Vic2NeighborProvince;
+}
+
+
+std::optional<int> getHoI4ProvinceNumber(int Vic2ProvinceNum, const Mappers::ProvinceMapper& provinceMapper)
+{
+	const auto& HoI4ProvinceNumbers = provinceMapper.getVic2ToHoI4ProvinceMapping(Vic2ProvinceNum);
+	if (HoI4ProvinceNumbers.empty())
+	{
+		return std::nullopt;
+	}
+
+	return HoI4ProvinceNumbers[0];
+}
+
+
+bool HoI4ProvinceNumbersAreValid(std::optional<int> firstNumber, std::optional<int> secondNumber)
+{
+	return firstNumber && secondNumber && *firstNumber == *secondNumber;
+}
+
+
+std::optional<std::vector<int>> findPath(int startProvince,
+	 int endProvince,
+	 const Maps::MapData& HoI4MapData,
+	 const Maps::ProvinceDefinitions& HoI4ProvinceDefinitions)
+{
+	std::vector<std::vector<int>> possibleRailwayPaths{{startProvince}};
+	std::set reachedProvinces{startProvince};
+
+	while (!possibleRailwayPaths.empty())
+	{
+		std::vector<int> possibleRailwayPath = possibleRailwayPaths[0];
+		if (possibleRailwayPath[possibleRailwayPath.size() - 1] == endProvince)
+		{
+			break;
+		}
+
+		for (const auto& neighborNumber: HoI4MapData.getNeighbors(possibleRailwayPath[possibleRailwayPath.size() - 1]))
```

Comment:
Do you have any geographical data you could plug in here and make this a proper A* instead? You'll slaughter the conversion if you do pathfinding like this, in onions without a heuristic function. Especially without caching.

Literally anything can be used, distance between 2 provinces on provinces.bmp for example.

Reply:
Surprisingly it takes < 1 sec to do it all. I think because its just neighbor to neighbor so the paths aren't long. But when I do improved routing I'll see how that goes and consider a proper A*.